### PR TITLE
(1) Testcases Editor: Data Integration / Organization of code (2) Problem Context: markProblemsDirty

### PR DESCRIPTION
**Pull request by Johnny850807**
(1) Testcases Editor: Data Integration / Organization of code (2) Problem Context: markProblemsDirty

### CODE REVIEW

File: src/app/problem/editor/right/testcases/usecase.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        const nextId = testcases ? Math.max(...testcases.map(t => t.id)) + 1 : 0;
+        console.log(`Next testcase's id: ${nextId}`);
+        setTestcases([...testcases, new Testcase({id: nextId})]);
+    };
+
+    const deleteTestcase = deletedId => {
+        setTestcases(testcasesList => testcasesList.filter(testcase => testcase.id !== deletedId));
+    };
+
+    return {testcases, setTestcases, addNewTestcase, deleteTestcase}
```
Comment (Wally5077) [inline src/app/problem/editor/right/testcases/usecase.js:18]: ```;```

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

File: src/app/problem/editor/right/testcases/Testcases.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import TestcaseEditor from "./TestcaseEditor";
+import {useProblemEditorContext} from "../../ProblemEditorContext";
+
+const TestCaseList = () => {
+    const {currentProblem} = useProblemEditorContext();
+    const {testcases, setTestcases, addNewTestcase, deleteTestcase} = useTestcaseList();
+
+    useEffect(() => {
+        setTestcases(currentProblem.testcases);
+    });
```
Comment (nyngwang) [inline src/app/problem/editor/right/testcases/Testcases.js:15]: Should it be `useEffect(() => { ... }, [currentProblem])`?

File: src/app/problem/editor/right/testcases/Testcases.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            <EditorButton
+                text={"Add New Test Case"}
+                buttonColor={"#1273BA"}
+                fontColor={"#FFFFFF"}
+                width={"100%"}
+                height={46}
+                onClick={addNewTestcase}
+            />
+        </div>
+    </>
```
Comment (nyngwang) [inline src/app/problem/editor/right/testcases/Testcases.js:33]: `;`

File: src/app/problem/editor/right/testcases/TestcaseEditor.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
+import React from "react";
```
Comment (nyngwang) [inline src/app/problem/editor/right/testcases/TestcaseEditor.js:1]: This line can be removed.

**Review by nyngwang — CHANGES_REQUESTED** [review]
Review done.

File: src/app/problem/editor/right/testcases/Testcases.js (hunk 2 above)
Comment (Johnny850807) [reply src/app/problem/editor/right/testcases/Testcases.js:15]: No, I passed the CI.

File: src/app/problem/editor/right/testcases/Testcases.js (hunk 3 above)
Comment (Johnny850807) [reply src/app/problem/editor/right/testcases/Testcases.js:33]: What?

**Review by nyngwang — APPROVED** [review]
LGTM!

**Review by Chaoyuuu — APPROVED** [review]
LGTM

**Pull request merged**